### PR TITLE
Support for functional claims via EqualityProof and EqualityProver

### PR DESCRIPTION
**Pull request by ehildenb**
This sets up a specialized prover for equality/functional proofs, which can be discharged much faster than reachability proofs. The motivation for this comes from the fact that we are currently using the reachability prover for KEVM equality proofs (via the "runLemma => doneLemma" style proofs), and it's quite a bit slower than it should be (many calls to RPC server instead of just one call to simplify).

- A simple test of an equality proof for IMP is added.
- The class `EqualityProof` is added, which stores as data all the components of the equality, as well as the simplified state of the equality. In addition, a convenience method `EqualityProof.from_claim` is provided for extracting `EqualityProof` from a `KClaim`.
- The class `EqualityProver` is added which holds an `EqualityProof` and is able to "advance" it by calling the simplifier on the underlying equality. If it makes progress, the simplified equality is stored as part of the proof state.

### CODE REVIEW

File: src/pyk/proof/equality.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+_LOGGER: Final = logging.getLogger(__name__)
+
+
+class EqualityProof(Proof):
+    lhs_body: KInner
+    rhs_body: KInner
+    sort: KSort
+    lhs_constraints: Iterable[KInner]
+    rhs_constraints: Iterable[KInner]
```
Comment (tothtamas28) [inline src/pyk/proof/equality.py:32]:
```suggestion
    lhs_constraints: tuple[KInner, ...]
    rhs_constraints: tuple[KInner, ...]
```

File: src/tests/integration/kcfg/test_imp.py [hunk 2]
```diff
@@ -495,3 +504,28 @@ def test_all_path_bmc_reachability_prove(
         )
 
         assert proof.status == proof_status
+
+    @pytest.mark.parametrize(
+        'test_id,spec_file,spec_module,claim_id,proof_status',
+        FUNC_PROVE_TEST_DATA,
+        ids=[test_id for test_id, *_ in FUNC_PROVE_TEST_DATA],
+    )
+    def test_functional_prove(
```
Comment (tothtamas28) [inline src/tests/integration/kcfg/test_imp.py:513]: Consider moving this to `tests/integration/proof/test_equality.py`.

**Review by tothtamas28 — APPROVED** [review]
Great feature! :rocket:

File: k-files/imp-simple-spec.k [hunk 3]
```diff
@@ -1,5 +1,12 @@
 requires "imp-verification.k"
 
+module IMP-FUNCTIONAL-SPEC
+    imports IMP-VERIFICATION
+
+    claim [concrete-addition]: 3 +Int 4 => 7
+
+endmodule
```
Comment (ana-pantilie) [inline k-files/imp-simple-spec.k:8]: What does the frontend transform this into?

Comment (ehildenb) [reply k-files/imp-simple-spec.k:8]: I don't think it sends it to kore, but I'm not sure. Here, we are using `KProve.get_claims`, which basically calls `kprove --dry-run --emit-json ...`. That is a KAST file, not a Kore file, which we read in directly, and then we use the RPC server to discharge it as follows:

A rule:

```
rule LHS => RHS requires REQ ensures ENS
```

Gets sent to the Haskell backend `simplify` endpoint as:

```
{ LHS #And REQ #Equals RHS #And ENS }
```

Comment (ehildenb) [reply k-files/imp-simple-spec.k:8]: So very little of the logic is happening in frontend, it's happening in Python, the frontend is just parsing things for us.

File: src/pyk/proof/equality.py (hunk 1 above)
Comment (ehildenb) [reply src/pyk/proof/equality.py:32]: Done.

**Comment by ehildenb** [comment]
This doesn't make #12 obsolete, because we still want to provide a CLI tool which users can call saying `pyk simplify ...` where they jjust provide a K term, and it produces the simplified variant of the term.

**Pull request merged**